File: he-lcmk.py
#                           जय श्री राम

import sys; import math; from collections import *
from functools import reduce
# sys.setrecursionlimit(10**6)

def get_ints(): return map(int, input().split())
def get_list(): return list(get_ints())
def printspx(*args): return print(*args, end="")
def printsp(*args): return print(*args, end=" ")

UGLYMOD = int(1e9)+7; SEXYMOD = 998244353; MAXN = int(1e6) + 3

#sys.stdin  = open("input.txt","r"); sys.stdout = open("output.txt","w")

Prime = [0] * MAXN
Prime[0] = Prime[1] = 1
for i in range(4, MAXN, 2):
    Prime[i] = 2
i = 3
while i*i <= MAXN:
    if not Prime[i]:
        for j in range(2*i, MAXN, i):
            if not Prime[j]:
                Prime[j] = i
    i += 2
# print(Prime)

def power(x, y, p): 
    res = 1;
    x = x % p;  
    while (y > 0):
        if (y & 1): 
            res = (res * x) % p; 
        y = y >> 1; # y = y/2 
        x = (x * x) % p; 
    return res; 

def getfacto(n):
    while n > 1:
        if Prime[n] == 0:
            facto[n] += 1
            lcm[n] = max(lcm[n], facto[n])
            n = 1
        else:
            dumb = Prime[n]
            while n % dumb == 0:
                facto[dumb] += 1
                n //= dumb
            lcm[dumb] = max(lcm[dumb], facto[dumb])
    return



for _testcases_ in range(int(input())):
    n, m, fuck = get_ints()
    li = get_list()

    lcm = defaultdict(int)

    for i in li:
        facto = defaultdict(int)
        getfacto(i)
    ans = 1
    for k, v in lcm.items():
        if v > 0:
            ans *= (power(k, (fuck*v), m)) % m
            ans %= m
    print(ans % m)



    # li = list(map(lambda x: power(x, k, m), li))
    # lcm = reduce(lambda x, y: ((x*y)//math.gcd(x, y))%m, li)
    # print(lcm % m)
    
    # for i in range(2, MAXN):
    #     if lcm[i]:
    #         # print(i, lcm[i])
    #         ans *= power(i, lcm[i]*k, m)
    # ans = power(lcm, k, m)
    # lcm = [0] * MAXN
    # for i in range(2, MAXN):
    #     if Prime[i]:
    #         totalFac = 0


            





'''
>>> COMMENT THE STDIN!! CHANGE ONLINE JUDGE !!
THE LOGIC AND APPROACH IS MINE @luctivud ( UDIT GUPTA )
Link may be copy-pasted here if it's taken from other source.
DO NOT PLAGIARISE.
>>> COMMENT THE STDIN!! CHANGE ONLINE JUDGE !!
'''

File: he-lcmk.cpp
/*     <<<  J A I ~ S H R E E ~ R A M  >>>     */

// Creator & Template by : Udit Gupta @luctivud


// #include "bits/stdc++.h"
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <sstream>
#include <queue>
#include <deque>
#include <bitset>
#include <iterator>
#include <list>
#include <stack>
#include <map>
#include <set>
#include <functional>
#include <numeric>
#include <utility>
#include <limits>
#include <time.h>
#include <math.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <assert.h>

using namespace std;

const long long int MOD_JOHAN {1000000007};
const long long int MOD_LIGHT {998244353};
const long long int MAXN_EYEPATCH = 1e6 + 1LL;
const long long int MAXN_FULLMETAL = 501LL;

typedef long long int lld;
typedef unsigned long long int llu;
typedef vector<int> v_int;
typedef vector<lld> v_lld;
typedef pair<int, int> p_int;
typedef pair<lld, lld> p_lld;


#define                   GREED_FOR_SPEED   ios_base::sync_with_stdio(false); cin.tie(0)
#define             TESTCASES_ARE_THERE()   _SC4N(Test0xcs); TotalT35tC4ses = Test0xcs; while(Test0xcs--)
#define                  mems(A77AY,V4LU)   memset((A77AY), V4LU, sizeof((A77AY)))
#define   FOR4WD(IT7T, ST47T, E9xD, St3P)   for(auto IT7T = ST47T; IT7T < E9xD; IT7T += St3P)
#define   FOR8WD(IT7T, ST47T, E9xD, St3P)   for(auto IT7T = ST47T; IT7T > E9xD; IT7T -= St3P)
#define  FOR4NGE(IT7T, ST47T, E9xD, St3P)   for(auto IT7T = ST47T;((St3P<0) ? IT7T>E9xD : IT7T<E9xD); IT7T += St3P)
#define                              endl   "\n"
#define            FORE4CH(V3CT07_, IT7T)   for(auto IT7T = V3CT07_.begin(); IT7T != V3CT07_.end(); IT7T++)
#define                    INP4T(V3CT07_)   for(auto &V3CT07____IT7T : V3CT07_) _SC4N(V3CT07____IT7T)
#define                                PI   3.1415926535897932384626433832795
#define                          INF1N1TY   1000000000000000005LL;
#define                    A47L(V3CT07_t)   V3CT07_t.begin(), V3CT07_t.end()
#define             L3N(V3CT07_x, Typ30F)   (Typ30F) V3CT07_x.size()
#define                            un_map   unordered_map
#define                            un_set   unordered_set


template<typename T> void _SC4N(T &x) { 
    x = 0; bool neg = 0; 
    register T c = getchar();
    if (c == '-') neg = 1, c = getchar();
    while ((c < 48) || (c > 57)) c = getchar();
    for ( ; c < 48||c > 57 ; c = getchar());
    for ( ; c > 47 && c < 58; c = getchar() ) x= (x << 3) + ( x << 1 ) + ( c & 15 );
    if (neg) x *= -1;
}


void printYesNo(bool Expr, int YNType = 1, string FirstChoice = "Ud", string SecondChoice = "Ud") { 
    if (YNType == 1) puts(Expr ? "YES" : "NO");
    else if (YNType == 2) puts(Expr ? "Yes" : "No");
    else if (YNType == 3) puts(Expr ? "yes" : "no");
    else if (YNType == 4) cout << (Expr ? FirstChoice : SecondChoice);
}

template <typename Type30F1> Type30F1 POWM0D(Type30F1 a, Type30F1 b, Type30F1 mod) {
    Type30F1 res = 1; a %= mod; assert(b>=0); for(; b ; b >>= 1) { if(b & 1) res = res * a % mod; a = a * a % mod; } return res;
}

vector <llu> prime(MAXN_EYEPATCH, 0);
map<llu, llu> mulcm;

void sieve() {
    prime[1] = prime[0] = 1;
    FOR4NGE(i, 4, MAXN_EYEPATCH, 2) {
        prime[i] = 2;
    }
    llu i = 3;
    while (i*i <= MAXN_EYEPATCH) {
        if (prime[i] == 0) {
            FOR4NGE(j, 2*i, MAXN_EYEPATCH, i) {
                if (prime[j] == 0) {
                    prime[j] = i;
                }
            }
        }
        i += 2;
    }
    return ;
}

void getfac(llu num) {
    map<llu, llu> temp ;
    while (num > 1) {
        if (prime[num] == 0) {
            temp[num] += 1;
            mulcm[num] = max(mulcm[num], temp[num]);
            num = 1;
        } else {
            llu foo = prime[num];
            while (num % foo == 0) {
                num /= foo;
                temp[foo] += 1;
            }
            mulcm[foo] = max(mulcm[foo], temp[foo]);
        }
    }
    return;
}


void solveEach(lld TestCaseNumber = 1) {
    llu n, m, k; _SC4N(n); _SC4N(m); _SC4N(k);
    vector<llu> arr(n); INP4T(arr);
    mulcm.clear();
    FORE4CH(arr, i) {
        getfac(*i);
    }
    llu ans = 1LL;
    for(auto it: mulcm) {
        ans *= (POWM0D(it.first, it.second * k, m)) % m;
        ans %= m;
    }
    cout << ans;
    cout << "\n"; 
    return;
}


signed main() {

    GREED_FOR_SPEED;

    lld Test0xcs = 0, TotalT35tC4ses = 1;

    sieve();

    #ifndef ONLINE_JUDGE
        freopen("input.txt", "r", stdin);
        freopen("output.txt", "w", stdout);
    #endif 

    TESTCASES_ARE_THERE()
        solveEach(TotalT35tC4ses - Test0xcs);
    return 0;
}

/*
THE LOGIC AND APPROACH IS MINE ( UDIT GUPTA )
SOME PARTS OF THE CODE MAY HAVE BEEN TAKEN FROM WEBSITES LIKE::
(I Own the code if no link is provided here or I may have missed mentioning it)ch|=spacelo &=_up change case
*/